File: matlab/test/unit/test_FGCalSine_overflowChk.m
%% test_FGCalSine_overflowChk.m - Unit test for overflowChk function
% Tests the overflowChk function with SAR ADC digital output data
%
% Output structure:
%   test_output/<data_set_name>/test_overflowChk/
%       overflowChk_matlab.png  - overflow check plot

close all; clc; clear;
warning("off");

%% Configuration
addpath('matlab/dout');
addpath('matlab/common');
addpath('matlab/test/unit');

inputDir = "test_data";
outputDir = "test_output";

% Test datasets - leave empty to auto-search
filesList = {};
filesList = autoSearchFiles(filesList, inputDir, 'dout_SAR_*.csv');

if ~isfolder(outputDir)
    mkdir(outputDir);
end

%% Test Loop
fprintf('=== test_FGCalSine_overflowChk.m ===\n');
fprintf('[Testing] %d datasets...\n\n', length(filesList));

for k = 1:length(filesList)
    currentFilename = filesList{k};
    dataFilePath = fullfile(inputDir, currentFilename);

    if ~isfile(dataFilePath)
        fprintf('[%d/%d] %s - NOT FOUND, skipping\n\n', k, length(filesList), currentFilename);
        continue;
    end
    fprintf('[%d/%d] [Processing] %s\n', k, length(filesList), currentFilename);

    read_data = readmatrix(dataFilePath);

    % Extract dataset name
    [~, datasetName, ~] = fileparts(currentFilename);
    titleString = replace(datasetName, '_', '\_');

    % Create output subfolder
    subFolder = fullfile(outputDir, datasetName, 'test_overflowChk');
    if ~isfolder(subFolder)
        mkdir(subFolder);
    end

    %% Run FGCalSine to get calibrated weights
    weights_cal = FGCalSine(read_data);

    %% Run overflowChk
    figure('Visible', 'off');
    overflowChk(read_data, weights_cal);
    title(titleString);

    % Save plot
    plotPath = fullfile(subFolder, 'overflowChk_matlab.png');
    saveas(gcf, plotPath);
    fprintf('  [Saved] %s\n\n', plotPath);
    close(gcf);
end

fprintf('[test_FGCalSine_overflowChk COMPLETE]\n');

File: adctoolbox/test/unit/test_FGCalSine_overflowChk.py
"""Test FGCalSine and overflowChk - Python version of test_FGCalSine_overflowChk.m"""

import numpy as np
import matplotlib.pyplot as plt
import warnings
from pathlib import Path

from adctoolbox.dout import FGCalSine, overflowChk


def test_fgcalsine_overflowchk():
    """Test FGCalSine and overflowChk with SAR ADC data."""

    # Suppress warnings for cleaner output
    warnings.filterwarnings('ignore')

    # Define input and output directories
    inputdir = Path("test_data")
    outputdir = Path("test_output")

    # File list
    files_list = [
        'dout_SAR_12b_weight_1.csv',
        'dout_SAR_12b_weight_2.csv',
        'dout_SAR_12b_weight_3.csv',
    ]

    # If files_list is empty, search for dout*.csv files
    if not files_list:
        search_pattern = "dout*.csv"
        matched_files = list(inputdir.glob(search_pattern))
        files_list = [f.name for f in matched_files]

    # Process each file
    for current_filename in files_list:
        data_filepath = inputdir / current_filename

        print(f"[Processing] {current_filename}")

        # Load data
        read_code = np.loadtxt(data_filepath, delimiter=',')
        if read_code.ndim == 1:
            read_code = read_code.reshape(-1, 1)

        print(f"  [Data shape] {read_code.shape}")

        # Call FGCalSine to get calibrated weights
        print(f"  [FGCalSine] Starting calibration...")
        weights_cal, offset, postCal, ideal, err, freqCal = FGCalSine(read_code)

        print(f"  [FGCalSine] Calibrated weights: {weights_cal}")
        print(f"  [FGCalSine] Offset: {offset:.6f}")
        print(f"  [FGCalSine] Calibrated frequency: {freqCal:.6f}")

        # Create figure for overflowChk
        fig = plt.figure(figsize=(10, 6))

        # Call overflowChk to visualize residue distribution
        overflowChk(read_code, weights_cal)

        # Set title (matching MATLAB format with underscores)
        name = current_filename.replace('.csv', '')
        title_string = name  # Keep underscores as-is (no escaping needed in Python)
        plt.title(title_string)

        # Save figure
        output_filepath = outputdir / f'{name}_overflowChk_python.png'
        plt.savefig(output_filepath, dpi=150, bbox_inches='tight')
        plt.close(fig)

        print(f"  [Saved image] -> [{output_filepath}]\n")

    # Re-enable warnings
    warnings.filterwarnings('default')

    # Close any remaining figures
    plt.close('all')


if __name__ == "__main__":
    test_fgcalsine_overflowchk()
